### PR TITLE
Use Content-Disposition for output file name in Webcmdlets

### PR DESCRIPTION
**Pull request by CarloToso**
# PR Summary

- Consider Content-Disposition in the autonaming of downloaded files
- Add handling of edge case in naming
- Add tests

## PR Context
Contributes to #11671 follow up to #19007

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/9788
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 1]
```diff
@@ -36,11 +36,34 @@ internal static class WebResponseHelper
...
+            {
+                string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
+                string pathAndQuery = response.RequestMessage.RequestUri.PathAndQuery;
+
+                if (!string.IsNullOrEmpty(contentDisposition)) 
+                {
+                    // Get file name from Content-Disposition header if present
+                    return Path.Join(_qualifiedOutFile, contentDisposition);
+                }
+                else if (pathAndQuery != "/")
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:50]: We can remove `else` here and below since `return`.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:50]: done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 2]
```diff
@@ -36,11 +36,33 @@ internal static class WebResponseHelper
         }
 
         internal static string GetOutFilePath(HttpResponseMessage response, string _qualifiedOutFile)
-        {
-            // Get file name from last segment of Uri
-            string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+        {            
+            if (Directory.Exists(_qualifiedOutFile))
+            {
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:41]: I suggest reduce nesting:
```c#
            if (!Directory.Exists(_qualifiedOutFile))
            {
                return _qualifiedOutFile;
            }

            string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
            ....

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: done

**Comment by CarloToso** [comment]
This PR works correctly, the new tests don't

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 3]
```diff
@@ -36,11 +36,31 @@ internal static class WebResponseHelper
         }
 
         internal static string GetOutFilePath(HttpResponseMessage response, string _qualifiedOutFile)
-        {
-            // Get file name from last segment of Uri
-            string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+        {            
+            if (!Directory.Exists(_qualifiedOutFile))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:40]: Eh, the parameter shouldn't have `_` prefix.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 4]
```diff
@@ -36,11 +36,31 @@ internal static class WebResponseHelper
...
+            {
+                return _qualifiedOutFile;
+            }
+
+            string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
+            string pathAndQuery = response.RequestMessage.RequestUri.PathAndQuery;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:46]: Please move to line 53.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 5]
```diff
@@ -748,6 +748,22 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
...
+
+        $uri = Get-WebListenerUrl -Test 'ResponseHeaders' -Query @{ "Content-Disposition" = $ContentDisposition }
+        $content = Invoke-WebRequest -Uri $uri
+        $outFile = Join-Path $TestDrive $ContentDisposition.FileName
+
+        # ensure the file does not exist
+        Remove-Item -Force -ErrorAction Ignore -Path $outFile
+        Invoke-WebRequest -Uri $uri -OutFile $TestDrive
+
+        Test-Path $outFile | Should -Be $true
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:763]:
```suggestion
        Test-Path $outFile | Should -BeTrue
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 6]
```diff
@@ -2722,6 +2738,22 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
...
+
+        $uri = Get-WebListenerUrl -Test 'ResponseHeaders' -Query @{ "Content-Disposition" = $ContentDisposition }
+        $content = Invoke-WebRequest -Uri $uri
+        $outFile = Join-Path $TestDrive $ContentDisposition.FileName
+
+        # ensure the file does not exist
+        Remove-Item -Force -ErrorAction Ignore -Path $outFile
+        Invoke-RestMethod -Uri $uri -OutFile $TestDrive
+
+        Test-Path $outFile | Should -Be $true
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:2753]:
```suggestion
        Test-Path $outFile | Should -BeTrue
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs (hunk 3 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:40]: I know, but I followed the convention used in `WebRequestPSCmdlet.Common.cs` when I cached QualifiedOutFile in _qualifiedOutFile

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:40]: It is not field here, it is local parameter.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 7]
```diff
@@ -36,11 +36,30 @@ internal static class WebResponseHelper
...
+
+            string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
+
+            if (!string.IsNullOrEmpty(contentDisposition)) 
+            {
+                // Get file name from Content-Disposition header if present
+                return Path.Join(_qualifiedOutFile, contentDisposition);
+            }
+
+            if (response.RequestMessage.RequestUri.PathAndQuery != "/")
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:53]: Could you please clarify why we need PathAndQuery. Maybe add a comment in the code.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:53]: We need it to handle this case:
```ps1
Invoke-WebRequest https://www.google.com/ -OutFile .\  #PathAndQuery == "/"
#--> OutFile: www_google_com
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 8]
```diff
@@ -35,12 +35,31 @@ internal static class WebResponseHelper
...
+            {
+                string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+
+                // Get file name from last segment of Uri
+                return Path.Join(qualifiedOutFile, lastUriSegment);
+            }
 
-            return Directory.Exists(_qualifiedOutFile) ? Path.Join(_qualifiedOutFile, lastUriSegment) : _qualifiedOutFile;
+            // File name not found use sanitized Host name instead
+            return Path.Join(qualifiedOutFile, response.RequestMessage.RequestUri.Host.Replace('.', '_'));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:62]: I want we explicitly confirm this fallback.
/cc @SteveL-MSFT @mklement0

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 9]
```diff
@@ -35,12 +35,31 @@ internal static class WebResponseHelper
             return headers;
         }
 
-        internal static string GetOutFilePath(HttpResponseMessage response, string _qualifiedOutFile)
-        {
-            // Get file name from last segment of Uri
-            string lastUriSegment = System.Net.WebUtility.UrlDecode(response.RequestMessage.RequestUri.Segments[^1]);
+        internal static string GetOutFilePath(HttpResponseMessage response, string qualifiedOutFile)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:38]: We should be more smart - check File.Exists for result file name and if there is a conflict add `(1)` or `(2)` and so on.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs (hunk 8 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:62]: I compared with curl.
Curl simply fails autonaming if the url has no last segment-
Cattura4

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs (hunk 9 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:38]: curl doesn't check for existing files and just overwrites

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs [hunk 10]
```diff
@@ -35,12 +35,31 @@ internal static class WebResponseHelper
...
+        {            
+            if (!Directory.Exists(qualifiedOutFile))
+            {
+                return qualifiedOutFile;
+            }
+
+            string contentDisposition = response.Content.Headers.ContentDisposition?.FileNameStar ?? response.Content.Headers.ContentDisposition?.FileName;
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:45]: Per https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#directives the value may have path information and should be stripped.  It may or may not be enclosed in quotes.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs (hunk 10 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:45]: The guideline says that "filename" is usually default name for "Save As". This force we think that we go in wrong direction trying to make the cmdlet too smart. We don't really know what the user wants in a particular scenario. Therefore, I propose follow @mklement0 's proposal  and introduce a new parameter AutoFileName as enum:
- Default. OutFile works as usually and requires exact path and file name.
- URL.  OutFile is a directory and file name is generated based on URL. Throw like curl if no last segment.
- Form. OutFile is a directory and file name is generated based on Content-Disposition. Throw if no Content-Disposition or filename in Content-Disposition.
This way we allow the user to do exactly what he needs to do.

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:45]: I'd prefer if we left the decision to the cmdlet, if the user wants to choose a name for the file he just has to write it.

**Comment by CarloToso** [comment]
@SteveL-MSFT I followed your suggestions

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 11]
```diff
@@ -2722,6 +2738,22 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
         Get-Item $outFile | Select-Object -ExpandProperty Length | Should -Be $content.Content.Length
     }
 
+    It "Invoke-RestMethod -OutFile folder Downloads the file and names it as the Content-Disposition header if present." {
+        $ContentDisposition = [System.Net.Http.Headers.ContentDispositionHeaderValue]::new("attachment")
+        $ContentDisposition.FileName = 'DownloadedFile.txt'
+
+        $uri = Get-WebListenerUrl -Test 'ResponseHeaders' -Query @{ "Content-Disposition" = $ContentDisposition }
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:2745]: should add a test variation that includes full path and quotes

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (hunk 11 above)
Comment (CarloToso) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:2763]: @SteveL-MSFT done

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs (hunk 10 above)
Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs:45]: @SteveL-MSFT what should be the best course of action?

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: @SteveL-MSFT any news on AutoFileName?

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebResponseHelper.CoreClr.cs]: @CarloToso sorry for the delay, I'll queue this up for discussion in Cmdlets WG

**Comment by SteveL-MSFT** [comment]
The @PowerShell/wg-powershell-cmdlets  reviewed this.  It seems that we should have a -OutFolder parameter that determines the filename in order: Content-Disposition header (if exists, removing path and quotes), last segment of URL (if exists), otherwise fail.  The cmdlet should fail fast if the filename cannot be determined when this switch is used and not after the transfer has completed.  It should be a runtime check to be mutually exclusive from -OutFile instead of making the parametersets more complex.

**Comment by CarloToso** [comment]
TODO in follow up PR: add sanitation such as https://github.com/PowerShell/PowerShell/issues/11671#issuecomment-1689159877

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 12]
```diff
@@ -896,27 +903,26 @@ internal virtual void ValidateParameters()
...
+                ErrorRecord error = GetValidationError(WebCmdletStrings.PassThruOutFileMissing, "WebCmdletPassThruOutFileMissingException");
                 ThrowTerminatingError(error);
             }
 
-            _qualifiedOutFile = ShouldSaveToOutFile ? QualifiedOutFile : null;
-
-            // OutFile must not be a directory to use Resume.
-            if (Resume.IsPresent && Directory.Exists(_qualifiedOutFile))
+            // Resume requires OutFile and can't be used with OutFolder..
+            if (Resume.IsPresent && OutFile is null)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:919]: Why Resume can't be used with OutFolder? I'd expect we get the same file name from the request.

And typo.
```suggestion
            // Resume requires OutFile and can't be used with OutFolder.
            if (Resume.IsPresent && OutFile is null)
```

Comment (CarloToso) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: It can't be used because the filename for OutFolder is calculated after the response and Resume is before the request

**Comment by CarloToso** [comment]
Unrelated test failures

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs (hunk 12 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: Maybe not for the PR. We could make first request with range 0-0 (first byte) and see if we got the filename. It is not expensive.

**Comment by mklement0** [comment]
@SteveL-MSFT, the [suggested logic](https://github.com/PowerShell/PowerShell/pull/19385#issuecomment-1708771429) in the WG decision makes sense, but introducing a new parameter `-OutFolder`, alongside `-OutFile` would be confusing, given that since v7.4.0 `-OutFile` _already accepts a directory path_ (in which case it uses the ultimate target URL's last segment as the file name).

Wouldn't it make more sense to add the functionality to `-OutFile` when given a _directory_ path? (Technically, that would be a breaking change from 7.4.0, however)

**Comment by SteveL-MSFT** [comment]
@mklement0 maybe it would be better to have `-OutPath` with `-OutFile` alias.

**Comment by mklement0** [comment]
@ImportTaste:

I can see the appeal of the additional `-ContentDisposition` switch allowing you to explicitly choose between using the content-disposition filename and the last-ultimate-URL-segment filename when a _directory_ path is passed to `-OutFile` / `-OutPath`.

But to me the question is whether this level of control is needed in practice; perhaps the suggested _implied_ logic (use content-disposition filename, if present, fall back to the last URL segment) is enough? The appeal is that you'll get the server-suggested filename by default, if available.

That said, as noted, this would technically be a breaking change to the current `-OutFile` behavior; if that is a concern, an additional parameter is a must.
